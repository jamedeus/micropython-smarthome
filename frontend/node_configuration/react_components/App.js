import React from 'react';
import InstanceCard from './InstanceCard';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import './Animation.css';


// Takes object and key prefix, returns all keys that begin with prefix
function filterObject(obj, prefix) {
    return Object.entries(obj).reduce((acc, [key, value]) => {
        if (key.startsWith(prefix)) {
            acc[key] = value;
        }
        return acc;
    }, {});
};


class App extends React.Component {
    constructor(props) {
        super(props);
        // Default state object if not received from context
        this.state = {
            metadata: {},
            wifi: {}
        };
    }

    componentDidMount() {
        // Overwrite state with config received from context (if present)
        const config = JSON.parse(document.getElementById("config").textContent);
        console.log(config)
        if (config) {
            this.setState({ ...config });
        }
    }

    logState = () => {
        console.log(this.state);
    };

    // Handler for add device and add sensor buttons
    addInstance = (category) => {
        this.setState(prevState => {
            // Get index of new instance
            const index = Object.keys(filterObject(prevState, category)).length + 1;

            // Add key to state object with empty config
            // Will be populated with metadata template when user selects type
            prevState[`${category}${index}`] = {}

            return prevState;
        });
    };

    // Handler for delete button on device and sensor cards
    deleteInstance = (id) => {
        this.setState(prevState => {
            // Remove deleted instance from state object
            delete prevState[id]
            // Decrement all subsequent IDs in state object to prevent gaps in index
            return update_ids(id, prevState);
        });
    };

    // Handler for type select dropdown in device and sensor cards
    changeInstanceType = (id, category, event) => {
        this.setState(prevState => {
            // Get config template for selected type from metadata object
            const template = metadata[`${category}s`][event.target.value]['config_template'];
            // Overwrite section in state object with config template
            prevState[id] = template;
            return prevState;
        })
    };

    // Handler for all inputs inside device and sensor cards
    // Takes device/sensor ID, param, and value; updates state object
    // TODO this has issues for components with multiple inputs (on/off paths, thermostat)
    handleInputChange = (id, param, value) => {
        console.log(param)
        this.setState(prevState => {
            prevState[id][param] = value;
            return prevState;
        });
    };

    // Render 2 column layout with device and sensor cards
    renderLayout = () => {
        const deviceEntries = (
            <TransitionGroup className="device-container">
                {Object.entries(this.state)
                .filter(([key, _]) => key.startsWith('device'))
                .map(([key, config]) => (
                    <CSSTransition key={key} timeout={500} classNames="fade">
                        <InstanceCard
                            key={key}
                            id={key}
                            category="device"
                            config={config}
                            onDelete={() => this.deleteInstance(key)}
                            onInputChange={(paramName, value) => this.handleInputChange(key, paramName, value)}
                            onTypeChange={(event) => this.changeInstanceType(key, "device", event)}
                        />
                    </CSSTransition>
                ))}
            </TransitionGroup>
        );

        const sensorEntries = (
            <TransitionGroup className="sensor-container">
                {Object.entries(this.state)
                .filter(([key, _]) => key.startsWith('sensor'))
                .map(([key, config]) => (
                    <CSSTransition key={key} timeout={500} classNames="fade">
                        <InstanceCard
                            key={key}
                            id={key}
                            category="sensor"
                            config={config}
                            onDelete={() => this.deleteInstance(key)}
                            onInputChange={(paramName, value) => this.handleInputChange(key, paramName, value)}
                            onTypeChange={(event) => this.changeInstanceType(key, "sensor", event)}
                        />
                    </CSSTransition>
                ))}
            </TransitionGroup>
        );

        return (
            <>
                <h1 className="text-center pt-3 pb-4">TEST</h1>

                <div id="page1" className="d-flex flex-column h-100">
                    <div className="row mt-3">
                        <div id="sensors" className="col-sm">
                            <h2 className="text-center">Add Sensors</h2>
                            {sensorEntries}
                            <div id="add_sensor" className="text-center position-relative">
                                <button className="btn-secondary btn mb-3" onClick={() => this.addInstance('sensor')}>Add Sensor</button>
                            </div>
                        </div>
                        <div id="devices" className="col-sm">
                            <h2 className="text-center">Add Devices</h2>
                            {deviceEntries}
                            <div id="add_device" className="text-center position-relative">
                                <button className="btn-secondary btn mb-3" onClick={() => this.addInstance('device')}>Add Device</button>
                            </div>
                        </div>
                    </div>
                </div>

                <div className="bottom">
                    <button className="log-button" onClick={() => this.logState()}>Log State</button>
                </div>
            </>
        );
    }

    render() {
        return this.renderLayout();
    }
}


// Called by deleteInstance, decrements IDs of all subsequent instances to prevent gaps
// Example: If device2 is deleted, device3 becomes device2, device4 becomes device3, etc
function update_ids(target, state) {
    // Get category (device or sensor) and index of removed instance
    const category = target.replace(/[0-9]/g, '');
    const index = target.replace(/[a-zA-Z]/g, '');

    // Get list of all instances in same category
    var instances = filterObject(state, category);

    // If target is device get list of sensors (used to update target IDs)
    if (category === 'device') {
        var sensors = filterObject(state, 'sensor');
        // Remove device from all sensor targets
        for (const sensor in sensors) {
            state[sensor]['targets'] = state[sensor]['targets'].filter(item => item !== target);
        };
    };

    // Iterate all instances in category starting from the removed instance index
    for (let i=parseInt(index); i<Object.entries(instances).length+1; i++) {
        // Removed index now available, decrement next index by 1
        const new_id = `${category}${i}`;
        const old_id = `${category}${i+1}`;
        state[new_id] = JSON.parse(JSON.stringify(state[old_id]));
        delete state[old_id];

        // Decrement device index in sensor targets lists to match above
        if (category === 'device') {
            for (const sensor in sensors) {
                if (state[sensor]['targets'].includes(old_id)) {
                    state[sensor]['targets'] = state[sensor]['targets'].filter(item => item !== old_id);
                    state[sensor]['targets'].push(new_id);
                };
            };
        };
    }

    return state;
};


export default App;
