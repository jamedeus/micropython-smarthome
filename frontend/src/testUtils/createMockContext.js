// Creates DOM elements containing JSON (mocks django json_script filter)
// Takes context id (arg passed to parse_dom_context) and json contents
const createMockContext = (id, json) => {
    const mockContext = document.createElement('div');
    mockContext.id = id;
    mockContext.textContent = JSON.stringify(json);
    document.body.appendChild(mockContext);
};

export default createMockContext;
