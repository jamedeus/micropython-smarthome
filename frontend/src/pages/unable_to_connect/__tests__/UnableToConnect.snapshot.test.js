import React from 'react';
import App from '../UnableToConnect';
import createMockContext from 'src/testUtils/createMockContext';

describe('UnableToConnect', () => {
    it('matches snapshot', () => {
        // Create mock state object
        createMockContext('target_node', 'Bedroom');

        // Render App, confirm matches snapshot
        const app = render(<App />);
        expect(app).toMatchSnapshot();
    });
});
