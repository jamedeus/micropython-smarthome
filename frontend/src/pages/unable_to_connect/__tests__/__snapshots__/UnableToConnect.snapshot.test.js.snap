// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnableToConnect matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="target_node"
    >
      "Bedroom"
    </div>
    <div>
      <div
        class="d-flex flex-column vh-100"
      >
        <div
          class="d-flex justify-content-between"
        >
          <button
            class="my-auto btn btn-adaptive"
            type="button"
          >
            <i
              class="bi-chevron-left"
            />
          </button>
          <h1
            class="my-3"
          >
            Connection Failed
          </h1>
          <button
            class="my-auto btn btn-adaptive"
            disabled=""
            type="button"
          >
            <i
              class="bi-list"
            />
          </button>
        </div>
        <div
          class="d-flex flex-column h-100 justify-content-evenly my-auto"
        >
          <p
            class="text-center my-3"
          >
            This page will reload automatically when the node is available
          </p>
          <div
            class="loading-animation m-auto my-3 loading-animation-lg"
            id="spinner"
          >
            <div />
            <div />
            <div />
            <div />
          </div>
          <div
            class="d-flex flex-column"
          >
            <h2
              class="text-center mb-3"
            >
              Troubleshooting
            </h2>
            <ul
              class="mx-auto"
            >
              <li>
                Unplug the node, plug back in, make sure the red light is on
              </li>
              <li>
                If the blue light stays on 30+ seconds there is a network issue
              </li>
              <ul>
                <li>
                  Is your 2.4 GHz network available?
                </li>
                <li>
                  Can you connect to the internet?
                </li>
                <li>
                  Did the SSID or password change?
                </li>
              </ul>
            </ul>
          </div>
          <div
            class="d-flex mx-auto mb-3"
          >
            <button
              class="btn btn-success"
              type="button"
            >
              Back to Overview
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="d-flex flex-column vh-100"
    >
      <div
        class="d-flex justify-content-between"
      >
        <button
          class="my-auto btn btn-adaptive"
          type="button"
        >
          <i
            class="bi-chevron-left"
          />
        </button>
        <h1
          class="my-3"
        >
          Connection Failed
        </h1>
        <button
          class="my-auto btn btn-adaptive"
          disabled=""
          type="button"
        >
          <i
            class="bi-list"
          />
        </button>
      </div>
      <div
        class="d-flex flex-column h-100 justify-content-evenly my-auto"
      >
        <p
          class="text-center my-3"
        >
          This page will reload automatically when the node is available
        </p>
        <div
          class="loading-animation m-auto my-3 loading-animation-lg"
          id="spinner"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
        <div
          class="d-flex flex-column"
        >
          <h2
            class="text-center mb-3"
          >
            Troubleshooting
          </h2>
          <ul
            class="mx-auto"
          >
            <li>
              Unplug the node, plug back in, make sure the red light is on
            </li>
            <li>
              If the blue light stays on 30+ seconds there is a network issue
            </li>
            <ul>
              <li>
                Is your 2.4 GHz network available?
              </li>
              <li>
                Can you connect to the internet?
              </li>
              <li>
                Did the SSID or password change?
              </li>
            </ul>
          </ul>
        </div>
        <div
          class="d-flex mx-auto mb-3"
        >
          <button
            class="btn btn-success"
            type="button"
          >
            Back to Overview
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
